### PR TITLE
Fix ModelViolationError while parsing repo files

### PR DESCRIPTION
**Pull request by tomasfratrik**
This error occurs when repo file has invalid definition, specifically the 'name' entry of the config file.

After this, we get rid of the traceback error, and the following errors are shown
```
2024-06-17 13:35:00.974356 [ERROR] Actor: system_facts
Message: Invalid repository definition: invalid-repo in: /etc/yum.repos.d/invalid.repo: The value of "name" field is None, but this is not allowed
Summary:
    Hint: For more directions on how to resolve the issue, see: https://access.redhat.com/solutions/6969001.
2024-06-17 13:35:01.545642 [ERROR] Actor: repositories_blacklist
Message: Actor didn't receive required messages: RepositoriesFact
```

Jira: RHEL-19249

### CODE REVIEW

File: repos/system_upgrade/common/actors/systemfacts/libraries/systemfacts.py
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/systemfacts/libraries/systemfacts.py:233]: There is a problem with this approach, if the `InvalidRepoDefinition` exception is raised and not caught and handled the way it's handled here, there will be no message (and possibly no other info besides traceback, not sure how the kwargs would be displayed).

You could move the message definition into the InvalidRepoDefinition constructor, or at least to the ctor call site. Also the message could contain the repoid and repofile directly in the string too, e.g. `Invalid repository definition: <repoid> in: <filepath>`. But it's good having them available as members still.

Personally I like the way it's done here: https://github.com/oamg/leapp-repository/blob/236483d5478ce845726ee05b7ca3e080105cdb80/repos/system_upgrade/el7toel8/libraries/isccfg.py#L11-L23
but keeping that up to you.

File: repos/system_upgrade/common/actors/systemfacts/libraries/systemfacts.py
Comment (matejmatuska) [inline repos/system_upgrade/common/actors/systemfacts/libraries/systemfacts.py:232]: The link in the hint seems to be incorrect (maybe a typo)?

Also the wording of the hint might be misleading for the users, visiting the link will to resolve the issue :), something like
`For more directions on how to resolve the issue, see <link>.` is more accurate.
Of course it depends on what is actually at the correct link.

**Review by matejmatuska — CHANGES_REQUESTED** [review]
The overall approach is ok, but there are some problems.

File: repos/system_upgrade/common/actors/systemfacts/libraries/systemfacts.py
Comment (tomasfratrik) [reply repos/system_upgrade/common/actors/systemfacts/libraries/systemfacts.py]: Changed the wording, and corrected the link.

File: repos/system_upgrade/common/actors/systemfacts/libraries/systemfacts.py
Comment (tomasfratrik) [reply repos/system_upgrade/common/actors/systemfacts/libraries/systemfacts.py]: It makes more sense! Put it there, and again tested it.

File: repos/system_upgrade/common/libraries/repofileutils.py
Comment (matejmatuska) [inline repos/system_upgrade/common/libraries/repofileutils.py:15]: I am not sure if there is any benefit to have the default arguments here, "None" would look weird in the message. I would leave this up to the caller. Wdyt?

**Review by matejmatuska — CHANGES_REQUESTED** [review]
Just one small suggestion :) code-wise ok, I am going to test it now

**Review by matejmatuska — CHANGES_REQUESTED** [review]
The PR description is outdated, from my testing on RHEL 8 the output looks like this, which is good:
```
2024-08-02 12:17:22.052435 [ERROR] Actor: system_facts
Message: Invalid repository definition: copr:copr.fedorainfracloud.org:group_oamg:leapp in: /etc/yum.repos.d/leapp-copr.repo: The value of "name" field is None, but this is not allowed
Summary:
    Hint: For more directions on how to resolve the issue, see: https://access.redhat.com/solutions/6969001.
2024-08-02 12:17:33.302747 [ERROR] Actor: repositories_blacklist
Message: Actor didn't receive required messages: RepositoriesFacts
```
After the PR desc is updated, squash the commits and briefly describe the changes in the final one, otherwise LGTM.

DO NOT MERGE YET - FOR NOW IT'S NOT TARGETED FOR THE UPCOMING RELEASE

File: repos/system_upgrade/common/libraries/repofileutils.py
Comment (tomasfratrik) [reply repos/system_upgrade/common/libraries/repofileutils.py:15]: You are right

**Comment by tomasfratrik** [comment]
The latest push squashes commits.

**Comment by fernflower** [comment]
Could you please add a unit test?

**Comment by matejmatuska** [comment]
> Could you please add a unit test?

Agree, I don't know how I forgot that, sorry.

**Comment by tomasfratrik** [comment]
> Could you please add a unit test?

I added it!

**Review by matejmatuska — APPROVED** [review]
Tests has been added and it works as expected. LGTM

**Comment by tomasfratrik** [comment]
The latest push squashed commits.

File: repos/system_upgrade/common/libraries/repofileutils.py
Comment (pirat89) [inline repos/system_upgrade/common/libraries/repofileutils.py:52]: @tomasfratrik  this is too generic and usually it's not good practice. Specify expected exceptions instead. We know about `ModelViolationError`. Add anything else what you found that could really happen. So far, we know from the history just about this one.

File: repos/system_upgrade/common/libraries/repofileutils.py
Comment (matejmatuska) [inline repos/system_upgrade/common/libraries/repofileutils.py:15]: Since this is public and in a shared library, add a docstring describing what is the exception used for.

Also update the docstring in the `parse_repofile` function to document that it can raise this exception and when.

**Review by matejmatuska — CHANGES_REQUESTED** [review]
@pirat89 pointed out that a unit test was added, but only for the system_facts actor. Tests are still missing for the `parse_repofile` and potentially `_parse_repository` functions.

Also docstring have to be included on public members in shared libraries (it would be okay in an actor lib).

File: repos/system_upgrade/common/libraries/repofileutils.py
Comment (tomasfratrik) [reply repos/system_upgrade/common/libraries/repofileutils.py]: Sorry forgot to reply, from testing, only ModelViolationError was present

File: repos/system_upgrade/common/libraries/repofileutils.py
Comment (tomasfratrik) [reply repos/system_upgrade/common/libraries/repofileutils.py]: Tests were added and docstrings were updated.

File: repos/system_upgrade/common/libraries/tests/test_repofileutils.py
Comment (matejmatuska) [inline repos/system_upgrade/common/libraries/tests/test_repofileutils.py:20]: I think it wouldn't hurt to add some more test cases.

Since you test the `_parse_repository_function` directly, you have more flexibility to test unexpected inputs. 
I would for example add a test with `'name'=None`, some test to make sure invalid type is caught (e.g. `'baseurl'=True`) and also a test with `repoid=None`. 

The inputs right now should always be a string, but this will catch errors if we change the function in the future or call it from somewhere else.

**Review by matejmatuska — CHANGES_REQUESTED** [review]
I would add a little more test cases, otherwise ok.

File: repos/system_upgrade/common/libraries/tests/test_repofileutils.py
Comment (tomasfratrik) [reply repos/system_upgrade/common/libraries/tests/test_repofileutils.py:29]: I added those test cases.

**Review by matejmatuska — APPROVED** [review]
Integration tests on RHEL7 are broken everywhere, not cause by these changes.

LGTM

**Comment by pirat89** [comment]
@matejmatuska seems that the non-els error we discussed has been caused just by infra issue. merging.

**Pull request merged**
